### ISSUE TITLE
confusing dead code warning inside macro annotated function

### ISSUE THREAD

**Issue by dvdsk** [rust-lang/rust#92554]
If I instrument an _unused function_ using the _tokio-tracing_ library I get a dead code warning that seems impossible.

The `instrument` procedural macro rewrites the _unused function_. Somewhere this causes the dead code warning for the function to not be reported. However dead code warnings from the function body, caused by the function not being used, are reported. As you miss the dead code warning regarding the function body you could get really confused how the function body code can be dead. 

The hiding of the dead code warning for the function is a known [issue](https://github.com/tokio-rs/tracing/issues/1366) of the tracing library. It has happens to another procedural macro:  [fn-error-context](https://github.com/andrewhickman/fn-error-context/issues/3) where it could be resolved.

Given the following [code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=9238421a4bfb74ad191bc552b0fdaff1):

```rust
use tracing; // 0.1.29

struct Test;
impl Test {
    #[tracing::instrument]
    fn never_used(var: i32) {
        const IS_USED: &str = "I get a dead code warning";

        if IS_USED == "" {
            println!("");
        }
    }
}

fn main() {
    let _test = Test;
}
```

The current output is:

```
warning: constant is never used: `IS_USED`
 --> src/main.rs:7:9
  |
7 |         const IS_USED: &str = "I get a dead code warning";
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: `playground` (bin "playground") generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 1.10s
```
_note how the unused code warning regarding the function argument is also hidden and only the dead code warning from the function body is emitted making this case extra confusing_

Ideally the output should look like:

```
warning: unused variable: `var`
 --> src/main.rs:5:19
  |
5 |     fn never_used(var: i32) {
  |                   ^^^ help: if this is intentional, prefix it with an underscore: `_var`
  |
  = note: `#[warn(unused_variables)]` on by default

warning: associated function is never used: `never_used`
 --> src/main.rs:5:8
  |
5 |     fn never_used(var: i32) {
  |        ^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: constant is never used: `IS_USED`
 --> src/main.rs:6:9
  |
6 |         const IS_USED: &str = "I get a dead code warning";
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

If `never_used()` is not a member of `Test` the code compiles without any warnings, which while sub optimal is less confusing (and maybe more desirable) then having the single unused code warning.

A next step in investigating this could be writing a minimal macro that still causes this behavior. Unfortunately I do not have time for that right now.

Compiler: `stable 1.57 ` but also confirmed to happen on `nightly`
